Add support for missing columns to OrcSelectiveRecordReader

oerling opened the pull request:
Adds adaptation to missing columns in OrcSelectiveRecordReader. Evaluates TupleDomainFilters against these and fails the whole split if any filter on a missing column fails with null.
Tests interaction of schema evolution with prefilled (e.g. partitioning) columns.

oerling commented:
This change is not intended.

 

 

 

 

From: Swapnil <notifications@github.com> 
Sent: Wednesday, September 18, 2019 5:04 PM
To: prestodb/presto <presto@noreply.github.com>
Cc: oerling <erling@xs4all.nl>; Author <author@noreply.github.com>
Subject: Re: [prestodb/presto] Default missing columns to nulls (#13417)

 

@swapsmagic commented on this pull request.

  _____  

In presto-hive/src/main/java/com/facebook/presto/hive/security/SqlStandardAccessControl.java <https://github.com/prestodb/presto/pull/13417#discussion_r325938946> :

> @@ -241,7 +241,7 @@ public void checkCanCreateViewWithSelectFromColumns(ConnectorTransactionHandle t
     @Override
     public void checkCanSetCatalogSessionProperty(ConnectorTransactionHandle transaction, ConnectorIdentity identity, String propertyName)
     {
-        if (!isAdmin(transaction, identity)) {
+        if (false && !isAdmin(transaction, identity)) {

Is this change intentional? As this means the method is not doing anything no matter what the input is.

—
You are receiving this because you authored the thread.
Reply to this email directly, view it on GitHub <https://github.com/prestodb/presto/pull/13417?email_source=notifications&email_token=AKPPPT7GX2BN5RRTMJ7KYE3QKK6XJA5CNFSM4IYDKJJ2YY3PNVWWK3TUL52HS4DFWFIHK3DMKJSXC5LFON2FEZLWNFSXPKTDN5WW2ZLOORPWSZGOCFGQILY#pullrequestreview-290260015> , or mute the thread <https://github.com/notifications/unsubscribe-auth/AKPPPT3ABDJYHCUJ2YLTX5LQKK6XJANCNFSM4IYDKJJQ> .

mbasmanova commented on presto-orc/src/main/java/com/facebook/presto/orc/OrcSelectiveRecordReader.java:
```diff
@@ -160,6 +164,20 @@ public OrcSelectiveRecordReader(
 
         requireNonNull(constantValues, "constantValues is null");
         this.constantValues = new Object[this.hiveColumnIndices.length];
+        for (Integer columnIndex : includedColumns.keySet()) {
```
Integer -> int

mbasmanova commented on presto-orc/src/main/java/com/facebook/presto/orc/OrcSelectiveRecordReader.java:
```diff
@@ -160,6 +164,20 @@ public OrcSelectiveRecordReader(
         requireNonNull(constantValues, "constantValues is null");
         this.constantValues = new Object[this.hiveColumnIndices.length];
+        for (Integer columnIndex : includedColumns.keySet()) {
+            if (!presentColumns.contains(columnIndex)) {
```
use isColumnPresent method from the base class: if (!isColumnPresent(columnIndex)) {

mbasmanova commented on presto-orc/src/main/java/com/facebook/presto/orc/OrcSelectiveRecordReader.java:
```diff
@@ -160,6 +164,20 @@ public OrcSelectiveRecordReader(
         this.constantValues = new Object[this.hiveColumnIndices.length];
+        for (Integer columnIndex : includedColumns.keySet()) {
+            if (!presentColumns.contains(columnIndex)) {
+                // A Any filter except is null on a missing column
```
typo: "A Any" -> "Any"

mbasmanova commented on presto-orc/src/main/java/com/facebook/presto/orc/OrcSelectiveRecordReader.java:
```diff
@@ -160,6 +164,20 @@ public OrcSelectiveRecordReader(
+            if (!presentColumns.contains(columnIndex)) {
+                // A Any filter except is null on a missing column
+                // fails the whole split. Filters on prefilled columns
+                // are already valuated, hence we only check filters
```
typo? valuated -> evaluated

mbasmanova commented on presto-orc/src/main/java/com/facebook/presto/orc/OrcSelectiveRecordReader.java:
```diff
@@ -171,6 +189,11 @@ public OrcSelectiveRecordReader(
 
+    private static boolean containsNonNullFilter(Map<Subfield, TupleDomainFilter> columnFilters)
+    {
+        return columnFilters != null && !columnFilters.entrySet().stream().map(Map.Entry::getValue).allMatch(filter -> filter.testNull());
```
- entrySet().stream().map(Map.Entry::getValue) can be simplified to values().stream()
- use TupleDomainFilter::testNull in allMatch
- should we check if filters are deterministic as well?

return columnFilters != null && !columnFilters.values().stream().allMatch(TupleDomainFilter::testNull);

mbasmanova commented on presto-orc/src/main/java/com/facebook/presto/orc/AbstractOrcRecordReader.java:
```diff
@@ -73,7 +73,7 @@
     private final long totalRowCount;
     private final long splitLength;
-    private final Set<Integer> presentColumns;
+    protected final Set<Integer> presentColumns;
```
this change is not needed; use isColumnPresent method instead

mbasmanova commented on presto-hive/src/test/java/com/facebook/presto/hive/TestHivePushdownFilterQueries.java:
```diff
@@ -514,6 +514,33 @@ public void testPathColumn()
         assertQuerySucceeds(session, "SELECT linenumber, \"$path\" FROM lineitem");
         assertQuerySucceeds(session, "SELECT linenumber, \"$path\" FROM lineitem WHERE length(\"$path\") % 2 = linenumber % 2");
     }
+    @Test
```
add empty line before @Test

oerling commented:
I addressed the comments. As for determinism, I don’t think we have nondeterministic filters at the place in question. They are made deeper down. So if you have (map[1] = 2 or map[1] is null), then this is true and map[1] = 1 is false and there is no matter of selective application of filters to some elements and not others. For filter functions a determinism check is definitely indicated but for TupleDomain ones I think we are fine without.

 

 

 

From: Maria Basmanova <notifications@github.com> 
Sent: Wednesday, September 18, 2019 7:41 PM
To: prestodb/presto <presto@noreply.github.com>
Cc: oerling <erling@xs4all.nl>; Mention <mention@noreply.github.com>
Subject: Re: [prestodb/presto] Default missing columns to nulls (#13417)

 

@mbasmanova requested changes on this pull request.

@oerling <https://github.com/oerling>  Overall looks good to me.

I don't believe first commit belongs to this PR. Adding flexibility to run tests in single-threaded mode is useful, but needs to be figured out separately.

Second commit looks great % some comments.

Commit message needs to be wrapped at 78 characters and needs to follow guidelines from https://chris.beams.io/posts/git-commit/ I'd update the commit message like so:

Add support for missing columns to OrcSelectiveRecordReader
 
Evaluate TupleDomainFilters against missing columns and fail the whole split 
if any deterministic filter fails on null. 
 
Add tests for interaction of schema evolution with prefilled (e.g. partitioning)
columns.
  _____  


In presto-orc/src/main/java/com/facebook/presto/orc/OrcSelectiveRecordReader.java <https://github.com/prestodb/presto/pull/13417#discussion_r325965777> :

> @@ -160,6 +164,20 @@ public OrcSelectiveRecordReader(
 
         requireNonNull(constantValues, "constantValues is null");
         this.constantValues = new Object[this.hiveColumnIndices.length];
+        for (Integer columnIndex : includedColumns.keySet()) {

Integer -> int

  _____  

In presto-orc/src/main/java/com/facebook/presto/orc/OrcSelectiveRecordReader.java <https://github.com/prestodb/presto/pull/13417#discussion_r325965845> :

> @@ -160,6 +164,20 @@ public OrcSelectiveRecordReader(
 
         requireNonNull(constantValues, "constantValues is null");
         this.constantValues = new Object[this.hiveColumnIndices.length];
+        for (Integer columnIndex : includedColumns.keySet()) {
+            if (!presentColumns.contains(columnIndex)) {

use isColumnPresent method from the base class: if (!isColumnPresent(columnIndex)) {

  _____  

In presto-orc/src/main/java/com/facebook/presto/orc/OrcSelectiveRecordReader.java <https://github.com/prestodb/presto/pull/13417#discussion_r325965937> :

> @@ -160,6 +164,20 @@ public OrcSelectiveRecordReader(
 
         requireNonNull(constantValues, "constantValues is null");
         this.constantValues = new Object[this.hiveColumnIndices.length];
+        for (Integer columnIndex : includedColumns.keySet()) {
+            if (!presentColumns.contains(columnIndex)) {
+                // A Any filter except is null on a missing column

typo: "A Any" -> "Any"

  _____  

In presto-orc/src/main/java/com/facebook/presto/orc/OrcSelectiveRecordReader.java <https://github.com/prestodb/presto/pull/13417#discussion_r325965970> :

> @@ -160,6 +164,20 @@ public OrcSelectiveRecordReader(
 
         requireNonNull(constantValues, "constantValues is null");
         this.constantValues = new Object[this.hiveColumnIndices.length];
+        for (Integer columnIndex : includedColumns.keySet()) {
+            if (!presentColumns.contains(columnIndex)) {
+                // A Any filter except is null on a missing column
+                // fails the whole split. Filters on prefilled columns
+                // are already valuated, hence we only check filters

typo? valuated -> evaluated

  _____  

In presto-orc/src/main/java/com/facebook/presto/orc/OrcSelectiveRecordReader.java <https://github.com/prestodb/presto/pull/13417#discussion_r325966143> :

> @@ -171,6 +189,11 @@ public OrcSelectiveRecordReader(
         streamReaderOrder = orderStreamReaders(columnTypes.keySet().stream().filter(index -> this.constantValues[index] == null).collect(toImmutableSet()), columnsWithFilters, filterFunctionInputs);
     }
 
+    private static boolean containsNonNullFilter(Map<Subfield, TupleDomainFilter> columnFilters)
+    {
+        return columnFilters != null && !columnFilters.entrySet().stream().map(Map.Entry::getValue).allMatch(filter -> filter.testNull());

*	entrySet().stream().map(Map.Entry::getValue) can be simplified to values().stream()
*	use TupleDomainFilter::testNull in allMatch
*	should we check if filters are deterministic as well?

return columnFilters != null && !columnFilters.values().stream().allMatch(TupleDomainFilter::testNull);

  _____  

In presto-orc/src/main/java/com/facebook/presto/orc/AbstractOrcRecordReader.java <https://github.com/prestodb/presto/pull/13417#discussion_r325966712> :

> @@ -73,7 +73,7 @@
 
     private final long totalRowCount;
     private final long splitLength;
-    private final Set<Integer> presentColumns;
+    protected final Set<Integer> presentColumns;

this change is not needed; use isColumnPresent method instead

  _____  

In presto-hive/src/test/java/com/facebook/presto/hive/TestHivePushdownFilterQueries.java <https://github.com/prestodb/presto/pull/13417#discussion_r325967439> :

> @@ -514,6 +514,33 @@ public void testPathColumn()
         assertQuerySucceeds(session, "SELECT linenumber, \"$path\" FROM lineitem");
         assertQuerySucceeds(session, "SELECT linenumber, \"$path\" FROM lineitem WHERE length(\"$path\") % 2 = linenumber % 2");
     }
+    @Test

add empty line before @test <https://github.com/test> 

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub <https://github.com/prestodb/presto/pull/13417?email_source=notifications&email_token=AKPPPT2TNIU7WSOFORCNHZLQKLRDDA5CNFSM4IYDKJJ2YY3PNVWWK3TUL52HS4DFWFIHK3DMKJSXC5LFON2FEZLWNFSXPKTDN5WW2ZLOORPWSZGOCFGYMJQ#pullrequestreview-290293286> , or mute the thread <https://github.com/notifications/unsubscribe-auth/AKPPPT2SBCKKUKXJJPQ7Q4DQKLRDDANCNFSM4IYDKJJQ> .

oerling commented:
Updated. What would you suggest about the removed settings? 

 

 

From: Maria Basmanova <notifications@github.com> 
Sent: Thursday, September 19, 2019 11:56 AM
To: prestodb/presto <presto@noreply.github.com>
Cc: oerling <erling@xs4all.nl>; Mention <mention@noreply.github.com>
Subject: Re: [prestodb/presto] Default missing columns to nulls (#13417)

 

@mbasmanova requested changes on this pull request.

@oerling <https://github.com/oerling> 

Add support for missing columns to OrcSelectiveRecordReader commit looks good except for a change in SqlStandardAccessControl.java.

Please, remove Single threaded env setting and allow user to set connector propertie… commit.

  _____  

In presto-hive/src/main/java/com/facebook/presto/hive/security/SqlStandardAccessControl.java <https://github.com/prestodb/presto/pull/13417#discussion_r326332269> :

> @@ -241,7 +241,7 @@ public void checkCanCreateViewWithSelectFromColumns(ConnectorTransactionHandle t
 
     @Override
 
     public void checkCanSetCatalogSessionProperty(ConnectorTransactionHandle transaction, ConnectorIdentity identity, String propertyName)
 
     {
 
-        if (false && !isAdmin(transaction, identity)) {
 
+        if (!isAdmin(transaction, identity)) {
 

remove this change

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub <https://github.com/prestodb/presto/pull/13417?email_source=notifications&email_token=AKPPPT6WPIMAIHHMD3NSWALQKPDNBA5CNFSM4IYDKJJ2YY3PNVWWK3TUL52HS4DFWFIHK3DMKJSXC5LFON2FEZLWNFSXPKTDN5WW2ZLOORPWSZGOCFKN67Q#pullrequestreview-290774910> , or mute the thread <https://github.com/notifications/unsubscribe-auth/AKPPPT7BRG5S2LQCEZPBJE3QKPDNBANCNFSM4IYDKJJQ> .
